### PR TITLE
Implement CI/CD workflows for Initia binary builds

### PR DESCRIPTION
**Pull request by LizJeong**
## Changes
- Add build-and-upload.yml: Main workflow coordinating build processes
  * Set up matrix strategy for multiple platforms (Linux, Darwin)
  * Trigger builds for amd64 and arm64 architectures
- Add build-darwin-amd64.yml and build-darwin-arm64.yml: Darwin-specific build workflows
  * Configure macOS environment and build process
  * Support amd64 and arm64 architectures separately
- Add build-linux-amd64.yml and build-linux-arm64.yml: Linux-specific build workflows
  * Set up Ubuntu environment and build process
  * Include steps for amd64 and arm64 builds separately
- Implement automatic uploads to GitHub Releases for all builds
- Add verification steps for built binaries

## Rationale and Improvement
This PR sets up comprehensive CI/CD pipelines for building Initia binaries across multiple platforms and architectures, ensuring consistent and automated releases. These workflows significantly improve the developer experience by automating the build and release process, reducing manual errors, and ensuring that binaries are available for all supported platforms and architectures.

## Additional Notes
- This PR improves the overall release process and should make it easier for users to access the latest Initia binaries for their specific platform.
- The automated nature of these workflows will help maintain consistency across releases and reduce the workload on maintainers.
- Important: Before applying these workflows, the `GH_RELEASE_TOKEN` secret variable needs to be set in the repository settings. This token is required for the workflows to function properly, especially for creating releases and uploading assets.

## Reviewer Considerations
- Please review the workflow files for any potential security concerns or optimizations.
- Confirm that the `GH_RELEASE_TOKEN` secret is properly referenced in the workflow files and that its permissions are correctly set.

### CODE REVIEW

File: .github/workflows/build-linux-amd64.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      arch:
+        required: true
+        type: string
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
```
Comment (etienne-napoleone) [inline .github/workflows/build-linux-amd64.yml:15]: actions/checkout has v4

File: Dockerfile.arm64 [hunk 2]
```diff
@@ -0,0 +1,34 @@
+FROM arm64v8/golang:1.22-bullseye AS go-builder
+
+# Install minimum necessary dependencies, build Cosmos SDK, remove packages
+RUN apt update
+RUN apt install -y curl git build-essential
+# debug: for live editing in the image
+RUN apt install -y vim
```
[the remark this answers is not in the document]
Comment (etienne-napoleone) [reply Dockerfile.arm64:7]: multi-stage build so shouldn't really impact final image size

**Review by etienne-napoleone — COMMENTED** [review]
Looking good! Trusting you on the specific platform builds working.
I just added some comments about general organization of the workflows and github token secret

File: Dockerfile.arm64 (hunk 2 above)
[the remark this answers is not in the document]
Comment (etienne-napoleone) [reply Dockerfile.arm64:7]: multi-stage build so shouldn't really impact final image size
best practice and maybe better perf but not -too- important

File: .github/workflows/build-darwin-arm64.yml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+          && make build \
+          && cd ./build \
+          && cp ~/go/pkg/mod/github.com/initia-labs/movevm@${MOVEVM_VERSION}/api/libmovevm.dylib ./ \
+          && cp ~/go/pkg/mod/github.com/initia-labs/movevm@${MOVEVM_VERSION}/api/libcompiler.dylib ./ \
+          && tar -czvf initia_"$VERSION"_Darwin_"$ARCH_NAME".tar.gz initiad libmovevm.dylib libcompiler.dylib \
+          && mv ./initia_"$VERSION"_Darwin_"$ARCH_NAME".tar.gz $GITHUB_WORKSPACE/ \
+          && rm -rf ./libmovevm.dylib ./libcompiler.dylib ./initiad
+  
+      - name: Release
+        uses: softprops/action-gh-release@v1
```
[the remark this answers is not in the document]
Comment (LizJeong) [reply .github/workflows/build-darwin-arm64.yml]: Applied in v2 version!

File: .github/workflows/build-darwin-arm64.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        with:
+          files: |
+            initia_${{ env.VERSION }}_Darwin_${{ env.ARCH_NAME }}.tar.gz
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_RELEASE_TOKEN }}
```
[the remark this answers is not in the document]
Comment (LizJeong) [reply .github/workflows/build-darwin-arm64.yml]: Testing showed that the secret was not needed in this situation, so it was removed.

File: .github/workflows/build-darwin-amd64.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+          && mv ./initia_"$VERSION"_Darwin_"$ARCH_NAME".tar.gz $GITHUB_WORKSPACE/ \
+          && rm -rf ./libmovevm.dylib ./libcompiler.dylib ./initiad
+  
+      - name: Release
+        uses: softprops/action-gh-release@v1
+        with:
+          files: |
+            initia_${{ env.VERSION }}_Darwin_${{ env.ARCH_NAME }}.tar.gz
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_RELEASE_TOKEN }}
```
[the remark this answers is not in the document]
Comment (LizJeong) [reply .github/workflows/build-darwin-amd64.yml]: Enter has been added to each step!

File: .github/workflows/build-darwin-arm64.yml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+          echo "MOVEVM_VERSION=${MOVEVM_VERSION}" >> $GITHUB_ENV        
+          echo "GOARCH=${{ inputs.arch }}" >> $GITHUB_ENV
+          echo "GOOS=darwin" >> $GITHUB_ENV
+          if [[ "${{ github.ref }}" == "refs/tags/"* ]]; then
+            VERSION=${GITHUB_REF#refs/tags/}
+          else
+            VERSION="v0.0.0-${GITHUB_SHA::8}"
+          fi
+          echo "VERSION=${VERSION}" >> $GITHUB_ENV
+          if [ "${{ inputs.arch }}" == "amd64" ]; then
```
[the remark this answers is not in the document]
Comment (LizJeong) [reply .github/workflows/build-darwin-arm64.yml]: Unnecessary conditional statements were removed and structured into concise code.

File: .github/workflows/build-linux-amd64.yml (hunk 1 above)
Comment (LizJeong) [reply .github/workflows/build-linux-amd64.yml]: Applied in v4 version!

File: .github/workflows/build-and-upload.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+on:
+  push:
+    tags:
+      - 'v*'
+
+jobs:
+  build-linux-amd64:
+    uses: ./.github/workflows/build-linux-amd64.yml
+    with:
+      arch: amd64
```
[the remark this answers is not in the document]
Comment (LizJeong) [reply .github/workflows/build-and-upload.yml]: Unnecessary conditional statements were removed and structured into concise code!

**Review by etienne-napoleone — APPROVED** [review]
lgtm!

if you can please add ending newlines to all files (only added to one)

**Review by etienne-napoleone — APPROVED** [review]
looks good, @Vritra4 feel free to merge

@LizJeong if the unneeded release github token secret was only created for this workflow, can you make sure to delete it?

**Review by Vritra4 — APPROVED** [review]

**Pull request merged**